### PR TITLE
Reduction of RAW' data size via adjustment of approx. cluster member data types

### PR DESCRIPTION
**Pull request by abaty**
This PR is aimed at significantly reducing the size of the RAW' data format which uses the Approximated Cluster class, as the current implementation is not fully optimized.  This is the subsequent PR for improving the format which is mentioned at the end of the first paragraph in the description of PR #38423.

The space savings is accomplished by changing the full-precision float class members of SiStripApproximatedCluster to integer data types having limited precision.  In the case of the cluster width and average charge, values are rounded to the nearest whole number.  For the barycenter, rounding to one decimal place was found to give a good tradeoff of tracking performance and space savings .  To avoid using a floating point number, the barycenter is multiplied by 10 and stored as a 16-bit integer, and is converted back when regenerating the cluster collection from approximated clusters.  This reduces the approximated cluster collection size by 37% over the current implementation, and the SiStrip detector payload by 45% compared to the conventional RAW format.  A small loss (~0.5%) of inclusive tracks was observed, but largely disappears when applying analysis-type track selections.  (More info on our tests at [1], which was discussed in our data format working group)

This PR also incorporates two suggestions from @makortel for cleaning up the approximated cluster code [2,3].

[1] https://twiki.cern.ch/twiki/pub/Main/HIDetectorReadout2020/ApproxClusters_integer10th_Aug25.pdf
[2] https://github.com/cms-sw/cmssw/commit/8871eff2433649ccd801afc52edc00bbcf2f54d2
[3] https://github.com/cms-sw/cmssw/commit/0492f1f0803339ca24396a0429fb8b0c51aae72b

#### PR validation:

This PR can be tested using WF 140.58 and checking step2 to verify the smaller approx. cluster size.  Step 3 can be used to check physics performance and tracking differences.

@icali @mandrenguyen

### CODE REVIEW

**Comment by mmusich** [comment]
test parameters:
   * workflows =  161, 140.58

**Comment by mandrenguyen** [comment]
@cmsbuild please test

File: DataFormats/SiStripCluster/src/SiStripApproximateCluster.cc [hunk 1]
```diff
@@ -1,7 +1,10 @@
 #include "DataFormats/SiStripCluster/interface/SiStripApproximateCluster.h"
+#include "DataFormats/SiStripCluster/interface/SiStripCluster.h"
+#include <algorithm>
```
Comment (mandrenguyen) [inline DataFormats/SiStripCluster/src/SiStripApproximateCluster.cc:3]: Where is this header being used?

Comment (abaty) [reply DataFormats/SiStripCluster/src/SiStripApproximateCluster.cc:3]: I believe it is needed for the std::max call on L17

Comment (mandrenguyen) [reply DataFormats/SiStripCluster/src/SiStripApproximateCluster.cc:3]: I think std::max comes from cmath, and that algorithm is not needed.

Comment (makortel) [reply DataFormats/SiStripCluster/src/SiStripApproximateCluster.cc:3]: According to cppreference `std::max()` comes from `algorithm`
https://en.cppreference.com/w/cpp/algorithm/max

**Comment by mandrenguyen** [comment]
@perrotta @qliphy  I want to draw your attention to this PR.  This is not the "technical bug fix" that I understand was discussed in the ORP and is planned to be backported from 12_6_X.  This PR effectively changes the compression level of the approximated clusters, and does have an effect on the relevant workflows (161 and 140.58)  It could be considered a bug-fix in the sense that the default compression level was doing nearly nothing.  In any case, this is needed for ion running, so we'll have to find a way to get it into 12_5_X. 
For reference, the authors have presented this change at the HIN PAG [meeting](https://indico.cern.ch/event/1193856/) just now, and agreed to present it at the RECO/AT meeting next week (I still need to create the agenda).

**Comment by perrotta** [comment]
> In any case, this is needed for ion running, so we'll have to find a way to get it into 12_5_X.

Technically it can be still be considered a bug fix, thus I don't see problems in backporting it to 12_5_X before 12_5_0.
Given the delays we have in building pre5, it is a pity that you cannot merge it in pre5 already, so that you can profit of the release validation of pre5. Perhaps it is still possible, provided @cms-sw/reconstruction-l2 signs it now...

**Comment by mandrenguyen** [comment]
+reconstruction
Local test show the expected decrease in the size of the approximatedClusters collection (30-40%), depending on the workflows, and the ~1% decrease in generalTracks, as described by the authors. 
Only the two heavy-ion workflows with rawprime are affected, aside from a single negligible difference in 
39496.0  and 39500.0 which cannot possible be related

**Comment by perrotta** [comment]
+1
- Merge for pre5

**Comment by slava77** [comment]
@abaty @mandrenguyen @icali 
was there some discussion/plan to add some information about the cluster shape. Perhaps a few bits could resolve quite a few issues downstream?
 connected to https://github.com/cms-sw/cmssw/pull/38423#issuecomment-1168996383

IIUC, the current strategy has a massive inefficiency for V0 reconstruction
Perhaps a separate issue for this is going to help to keep track of the progress.

**Comment by abaty** [comment]
@slava77 Currently we are working on another raised issue regarding the backwards compatibility of this strategy, so we have not had time to critically examine the feasibility/accuracy of storing cluster shape info at HLT.  We will look into this issue once the backwards compatibility is resolved.  For the upcoming HI run, we have at the moment disabled the cluster shape filter so the massive V0 inefficiency is not expected.  I agree that making a separate issue to track this for the future would be a good idea.

> was there some discussion/plan to add some information about the cluster shape. Perhaps a few bits could resolve quite a few issues downstream? connected to [#38423 (comment)](https://github.com/cms-sw/cmssw/pull/38423#issuecomment-1168996383)
> 
> IIUC, the current strategy has a massive inefficiency for V0 reconstruction (assuming that as in pp the cluster shape filter stays) Perhaps a separate issue for this is going to help to keep track of the progress.

**Pull request merged**